init.lua

```diff
@@ -48,11 +46,81 @@ function cutorch.createCudaHostTensor(...)
    else
       size = torch.LongTensor{...}
    end
+   return size
+end
 
+-- Creates a FloatTensor using the CudaHostAllocator.
+-- Accepts either a LongStorage or a sequence of numbers.
+function cutorch.createCudaHostTensor(...)
+   local size = longTensorSize(...)
    local storage = torch.FloatStorage(cutorch.CudaHostAllocator, size:prod())
    return torch.FloatTensor(storage, 1, size:storage())
 end
 
+-- Creates a CudaTensor using the CudaUVAAllocator.
+-- Accepts either a LongStorage or a sequence of numbers.
+function _createUVATensor(...)
+   local size = longTensorSize(...)
+   local storage = torch.FloatStorage(cutorch.CudaUVAAllocator, size:prod())
+   return torch.FloatTensor(storage)
+end
+
+function cutorch.createFloatUVATensor(...)
+   return _createUVATensor(...)
+end
+
+-- Creates a CudaTensor using the CudaUVAAllocator.
+-- Accepts either a LongStorage or a sequence of numbers.
+-- First creates a UVA backed FloatTensor and takes its pointer.
+function cutorch.createCudaUVATensor(...)
+   local ft = _createUVATensor(...)
+   local storage = torch.CudaStorage(
+      ft:storage():size(),
+      tonumber(torch.data(ft:storage(), true))
+   )
+   return torch.CudaTensor(storage)
+end
+
+-- UVA storage is a single memory location backed by virtual addressing.
+-- Converting between CPU / GPU tensor types is done by raw pointer passing.
+-- We only support FloatTensor, CudaTensor, Cuda -> float and float -> Cuda atm
+function cutorch.toFloatUVATensor(t)
+   if not torch.type(t):find('Tensor') then
+      error('Must use a tensor, got ' .. torch.type(t))
+   end
+   -- See THCAllocator.c for explanation about synchronization
+   cutorch.streamSynchronize(0)
```

Comment:
Why the `cutorch.streamSynchronize(0)`? Also I don't see any explanation in THCAllocator.c about synchronization